src/binxml/assemble.rs

```diff
@@ -33,144 +34,283 @@ pub fn parse_tokens<W: Write, T: BinXmlOutput<W>>(
+                            }
+                        };
+                    }
+                    BinXMLDeserializedTokens::RefValue(value) => {
+                        trace!("BinXMLDeserializedTokens::Value(value) - {:?}", value);
+                        match current_element.take() {
+                            // A string that is not inside any element, yield it
+                            None => match value {
+                                BinXmlValue::EvtXml => {
+                                    panic!("Call `expand_templates` before calling this function")
+                                }
+                                _ => {
+                                    model.push(XmlModel::Value(Cow::Borrowed(value)));
+                                }
+                            },
+                            // A string that is bound to an attribute
+                            Some(builder) => {
+                                current_element = Some(builder.attribute_value(Cow::Borrowed(value)));
+                            }
+                        };
+                    }
+                    BinXMLDeserializedTokens::CDATASection => {}
+                    BinXMLDeserializedTokens::CharRef => {}
+                    BinXMLDeserializedTokens::EntityRef(e) => unimplemented!("{}", &format!("{:?}", e)),
+                    BinXMLDeserializedTokens::PITarget => {}
+                    BinXMLDeserializedTokens::PIData => {}
+                    BinXMLDeserializedTokens::Substitution(_) => {
+                        panic!("Call `expand_templates` before calling this function")
+                    }
+                    BinXMLDeserializedTokens::EndOfStream => model.push(XmlModel::EndOfStream),
+                    BinXMLDeserializedTokens::StartOfStream => model.push(XmlModel::StartOfStream),
+                }
             }
-            BinXMLDeserializedTokens::EndOfStream => model.push(XmlModel::EndOfStream),
-            BinXMLDeserializedTokens::StartOfStream => model.push(XmlModel::StartOfStream),
         }
     }
     model
 }
 
 pub fn expand_templates(
     token_tree: Vec<BinXMLDeserializedTokens>,
-) -> Vec<BinXMLDeserializedTokens> {
+) -> Vec<Cow<BinXMLDeserializedTokens>> {
     // We can assume the new tree will be at least as big as the old one.
     let mut stack = Vec::with_capacity(token_tree.len());
 
-    fn _expand_templates<'c>(
-        token: BinXMLDeserializedTokens<'c>,
-        stack: &mut Vec<BinXMLDeserializedTokens<'c>>,
+    fn _expand_templates<'a>(
+        token: Cow<'a, BinXMLDeserializedTokens<'a>>,
+        stack: &mut Vec<Cow<'a, BinXMLDeserializedTokens<'a>>>,
     ) {
         match token {
-            BinXMLDeserializedTokens::Value(value) => match value {
-                BinXmlValue::BinXmlType(tokens) => {
-                    for token in tokens.into_iter() {
-                        _expand_templates(token, stack);
+            Cow::Owned(owned_token) => {
```

Comment:
Again, this is not very clean IMO. I’m not sure I entierly understand the flow here.